### DISCUSSION TITLE
Seeking Guidance on TetrahedronViscoelasticityFEMForceField in SofaViscoElastic Plugin

### DISCUSSION THREAD

**Discussion by zq-isgood** [sofa-framework/sofa#4311] [Develop in C++] [answered]
Hi.
I am seeking guidance from the community on the following inquiries:
1. How can I extract the stiffness matrix from the TetrahedronViscoelasticityFEMForceField in SofaViscoElastic?
2. While attempting to use the ConjugateGradientSolver from the sofaCaribou plugin to solve for CG.A() on TetrahedronViscoelasticityFEMForceField, I have encountered an issue where every element of the stiffness matrix appears as NaN. Interestingly, when I replace the force field with TetrahedronElasticForce, the corresponding values are present. Could someone kindly provide insights into the potential reasons for this discrepancy?
This is a portion of my code."
![2023-11-21 10-50-59 的屏幕截图](https://github.com/sofa-framework/sofa/assets/114601569/227466a7-45a4-4ce2-b1f9-e0092c682447)
![2023-11-21 10-51-07 的屏幕截图](https://github.com/sofa-framework/sofa/assets/114601569/948c6455-7aab-4897-8c7e-84f1fefa638d)
![2023-11-21 10-51-28 的屏幕截图](https://github.com/sofa-framework/sofa/assets/114601569/4e2f18e1-8ba9-46f8-900a-84642f2a083f)

**Comment by alxbilger** [comment, marked as answer]
Hey,

1. You should have a look on [this discussion](https://github.com/sofa-framework/sofa/discussions/3350#discussioncomment-3760896).
2. Why do you use a component in Caribou while the rest of your scene is not from Caribou? If you want a conjugate gradient, use the one from SOFA : `CGLinearSolver`. I don't think Caribou and not-Caribou components are able to interact.

**Reply by zq-isgood** [reply]
Hi,@alxbliger,
Thank you for your fast answer.  I also found that I could not use these two plug-ins together. Your example taught me another way to extract the stiffness matrix.
I asked the author of the SofaViscoElastic plugin about how to extract the stiffness matrix, and the author replied “the Stiffness Matrix Cannot be exported directly through the Python Bindings”
What other plug-ins have viscoelastic models and can derive the stiffness matrix?
>

**Reply by alxbilger** [reply]
I don't see any reason why it is not possible to extract the stiffness matrix. The function `addKToMatrix` is implemented in the `TetrahedronViscoelasticityFEMForceField` component, so it should contribute to the mechanical system. I suggest you try with the example I gave, and let me know.

**Reply by zq-isgood** [reply]
I changed the example of the plugin:creep_test.py. Here is the code and error
![2023-11-24 16-30-37 的屏幕截图](https://github.com/sofa-framework/sofa/assets/114601569/ab0d56b6-964b-4195-a16c-5d895ab300e7)
![2023-11-24 16-30-42 的屏幕截图](https://github.com/sofa-framework/sofa/assets/114601569/94c4657a-860a-45eb-9096-44efa3c0c427)
![2023-11-24 16-30-53 的屏幕截图](https://github.com/sofa-framework/sofa/assets/114601569/3958ce47-c29c-4c90-99c8-8bc7067e1fc5)
![2023-11-24 16-30-56 的屏幕截图](https://github.com/sofa-framework/sofa/assets/114601569/9e3f8e5d-1601-4dd6-acba-e7da138c7d2e)
![2023-11-24 16-31-14 的屏幕截图](https://github.com/sofa-framework/sofa/assets/114601569/519b5643-2ec5-479f-a6be-0eec162366b4)

**Reply by alxbilger** [reply]
I don't see `from Sofa import SofaLinearSolver` that is present in the provided example

**Comment by zq-isgood** [comment]
Thank you for your help Alex.I will mark your first answer as solution as it is the more relevant for future readers.After I added that sentence, the stiffness matrix can be obtained. The following is the result.
![2023-11-24 17-20-09 的屏幕截图](https://github.com/sofa-framework/sofa/assets/114601569/3846e60c-3806-40c9-9dbc-ba3fd8b07d20)
